### PR TITLE
Fix issuance request binding

### PR DESCRIPTION
**Pull request by chris-wood**
The gist of the request generation process is now:

1. Generate a blinded token request;
2. Generate a blinded key pair;
3. Encrypt the origin name, using (1) and (2) as AAD.
4. Generate a signature covering (1-3).

The signature is just Ed25519, and the blinded key pair is just a randomized Ed25519 key pair. 

cc @FredericJacobs

### CODE REVIEW

**Review by tfpauly — APPROVED** [review]
Document flow changes look good, and (in my non-expert opinion) the approach seems like an improvement.

File: draft-private-access-tokens.md [hunk 1]
```diff
@@ -780,28 +780,36 @@ certificate pinning, to mitigate the risk of channel compromise; see
...
+
+The Client then encrypts the origin name using ISSUER_KEY, producing
+`issuer_key_id` and `encrypted_origin_name` as described in {{encrypt-origin}}.
 
-The Client uses CLIENT_KEY and CLIENT_SECRET to generate proof of its request as
-described in {{client-stable-mapping}}, yielding output client_origin_index_blind and
-client_origin_index_request. The Client then constructs a Private Access Token
-request using client_origin_index_request, blinded_req, and origin information.
+Finally, the Client uses CLIENT_SECRET to produce `request_proof`
+as described in {{index-proof}}.
```
Comment (janaiyengar) [inline draft-private-access-tokens.md:799]: n00b question: what is the purpose of this proof?

File: draft-private-access-tokens.md [hunk 2]
```diff
@@ -1201,7 +1240,7 @@ key (CLIENT_KEY), Mediators complete the mapping computation as follows:
 In pseudocode, this is as follows:
 
 ~~~
-evaluated_key = DeserialiesElement(client_origin_index_result)
+evaluated_key = DeserialiesElement(index_result)
```
Comment (janaiyengar) [inline draft-private-access-tokens.md:1243]:
```suggestion
evaluated_key = DeserializeElement(index_result)
```

File: draft-private-access-tokens.md [hunk 3]
```diff
@@ -1080,49 +1092,71 @@ computing the mapping output.
...
+index_key = SerializeElement(blind_key)
+index_base = SerializeElement(blind_generator)
+index_request = concat(index_key, index_base)
+index_blind = SerializeScalar(blind)
+~~~
+
+### Client Index Proof {#index-proof}
+
+Given a client secret (CLIENT_SECRET), Clients produce proof of their request
+based on the following inputs defined in {{request-one}}: `token_key_id`,
```
Comment (janaiyengar) [inline draft-private-access-tokens.md:1129]: another n00b question: what is the purpose of this proof? From the rest of the construction, this is for the Issuer, but I don't understand what this proof is about.

**Review by janaiyengar — COMMENTED** [review]
I'm trying to understand this, so this isn't really a review. I've got just questions (and a nit) but I've probably got not much to add here.

File: draft-private-access-tokens.md (hunk 1 above)
Comment (chris-wood) [reply draft-private-access-tokens.md]: The new formulation should be much simpler - please have a look!

File: draft-private-access-tokens.md (hunk 3 above)
Comment (chris-wood) [reply draft-private-access-tokens.md]: See comment above =)

File: draft-private-access-tokens.md [hunk 4]
```diff
@@ -644,14 +647,10 @@ operation:
 
 - RSA Blind Signatures {{BLINDSIG}}, for issuing and constructing Tokens as described in {{redemption}}.
 
-- Prime Order Groups (POGs), for computing stable mappings between (Client, Origin) pairs. This
-  document uses notation described in {{!VOPRF=I-D.irtf-cfrg-voprf, Section 2.1}}, and, in particular,
-  the functions RandomScalar(), Generator(), SerializeScalar(), SerializeElement(), and HashToScalar().
-
-- Non-Interactive proof-of-knowledge (POK), as described in {{nizk-dl}}, for verifying correctness of Client requests.
+- Schnorr signatures, as described in {{!RFC8032}}, for verifying correctness of Client requests.
```
Comment (tfpauly) [inline draft-private-access-tokens.md:650]: Is this still Schnorr? Or is it ed25519?

Comment (chris-wood) [reply draft-private-access-tokens.md:650]: They are the same, but I'll fix =)

**Review by dvorak42 — APPROVED** [review]

**Review by smhendrickson — APPROVED** [review]
LGTM. This is a great simplification!

File: draft-private-access-tokens.md [hunk 5]
```diff
@@ -629,11 +629,14 @@ This section describes the Issuance protocol for a Client to request and receive
 a token from an Issuer. Token issuance involves a Client, Mediator, and Issuer,
 with the following steps:
 
-1. The Client sends a token request to the Mediator, encrypted using an Issuer-specific key
+1. The Client sends a token request containing a token request, encrypted origin
+   name, and one-time-use public key and signature to the Mediator
```
Comment (janaiyengar) [inline draft-private-access-tokens.md:633]: s/Mediator/Attester, here and below

File: draft-private-access-tokens.md [hunk 6]
```diff
@@ -824,21 +823,21 @@ object carrying ORIGIN_TOKEN_KEY.
 
 - "blinded_req" is the Nk-octet request defined above.
 
-- "index_request" is computed as described in {{index-request}}.
+- "request_key" is computed as described in {{index-request}}.
```
Comment (janaiyengar) [inline draft-private-access-tokens.md:826]:
```suggestion
- "request_key" is computed as described in {{request-key}}.
```

and fix it below as well

File: draft-private-access-tokens.md [hunk 7]
```diff
@@ -789,13 +788,13 @@ blinded_req, blind_inv = rsabssa_blind(ORIGIN_TOKEN_KEY, message)
...
-and blind `index_blind` as described in {{client-stable-mapping}}.
+The Client then uses CLIENT_KEY to generate its one-time-use request public
+key `request_key` and blind `request_key_blind` as described in {{client-stable-mapping}}.
 
 The Client then encrypts the origin name using ISSUER_KEY, producing
 `issuer_key_id` and `encrypted_origin_name` as described in {{encrypt-origin}}.
 
-Finally, the Client uses CLIENT_SECRET to produce `request_proof`
+Finally, the Client uses CLIENT_SECRET to produce `request_signature`
 as described in {{index-proof}}.
```
Comment (janaiyengar) [inline draft-private-access-tokens.md:798]:
```suggestion
as described in {{request-signature}}.
```

and fix below as well

File: draft-private-access-tokens.md [hunk 8]
```diff
@@ -869,9 +868,9 @@ the Client. If the key does not match, the Mediator rejects the request with an
 400 error. Note that Mediators need to be careful in cases of key rotation; see
 {{privacy-considerations}}.
 
-The Mediator finally checks to ensure that the AccessTokenRequest.index_request is valid
-for the given CLIENT_KEY; see {{nizk-dl}} for verification details. If the index is invalid,
-the Mediator rejects the request with an HTTP 400 error.
+The Mediator finally checks to ensure that the AccessTokenRequest.request_key is valid
+for the given CLIENT_KEY; see {{client-stable-mapping}} for verification details. If the
+index is invalid, the Mediator rejects the request with an HTTP 400 error.
```
Comment (janaiyengar) [inline draft-private-access-tokens.md:873]:
```suggestion
key is invalid, the Mediator rejects the request with an HTTP 400 error.
```

File: draft-private-access-tokens.md [hunk 9]
```diff
@@ -1090,210 +1089,154 @@ describes Issuer behavior for computing the mapping with its per-Origin secret,
...
 ### Client Behavior {#client-stable-mapping}
 
-This section describes the Client behavior for generating an index request and proof.
+This section describes the Client behavior for generating an one-time-use
+request key and signature. Clients provide their secret CLIENT_KEY as input
+to the request key generation step, and the rest of the token request inputs
+to the signature generation step.
 
-#### Index Request {#index-request}
+#### Request Key {#index-request}
```
Comment (janaiyengar) [inline draft-private-access-tokens.md:1116]:
```suggestion
#### Request Key {#request-key}
```

**Review by janaiyengar — COMMENTED** [review]
A few comments but this is much clearer, thank you!

**Comment by chris-wood** [comment]
Thanks, @janaiyengar! @tfpauly, shall we close this out and focus on the new doc?

**Comment by chris-wood** [comment]
Closing.

**Pull request closed without merging**